src/internal/m365/collection/drive/permission.go

```diff
@@ -445,6 +445,102 @@ func UpdateLinkShares(
 	return alreadyDeleted, nil
 }
 
+func filterUnavailableEntitiesInLinkShare(
```

Comment:
these two funcs could be merged pretty easily with generics.

```go
nc filterToExistingResourceEntities[T any](
 	ctx context.Context,
 	perms []T,
 	availableEntities AvailableEntities,
 	oldIDToNewID syncd.MapTo[string],
 ) []T {
```

Reply:
Hmm, maybe I am missing something, but I could not get it folded into one. They are slightly different enough to cause trouble.

Reply:
Ah, looks like you'd need to extend an interface.  Then it'd work.  Take it or leave it, no big deal.

```go
type idAndTyper interface {
  ID() string
  Type() metadata.GV2Type
}

func filterToExistingResourceEntities[T idAndTyper](
 	ctx context.Context,
 	orig []T,
 	availableEntities ResourceIDNames,
 	oldIDToNewID syncd.MapTo[string],
 ) []T {
 	if availableEntities.Users == nil || availableEntities.Groups == nil {
 		logger.Ctx(ctx).Info("no resources provided for filtering")
 		return orig
 	}

 	filtered := []T{}

 	for _, t := range orig {
 		available := false

 		switch t.Type() {
 		case metadata.GV2User:
 			_, ok := availableEntities.Users.NameOf(t.ID())
 			available = available || ok
 		case metadata.GV2Group:
 			_, ok := availableEntities.Groups.NameOf(t.ID())
 			available = available || ok
 		default:
 			// We only know about users and groups
 			// TODO: extend the check to other entity types
 			available = true
 		}

 		if available {
 			filtered = append(filtered, t)
 			continue
 		}

 		// If we have no entities, we can't restore the permission
 		// and so we have to mark it as not restored.
 		oldPermIDToNewID.Store(t.ID(), "")
 	}

 	return filtered
 }
```

Reply:
Yeah that works, but I think I'll skip it the primary reason being that the implementation within is also slightly different. Permissions has a single entry and link shares can have multiple entities. Don't think it will give a big advantage.